Add timezone dropdown

alexandercyu opened the pull request:
Adds a dropdown to the Temporal section that allows a user  to select a timezone.
The time selected in the calendar is preserved.  (If the user's data is at 3pm in California the return from the server will show 3pm and the timezone dropdown will show GMT-8).

Note that these timezones are GMT centered and the offsets are calculated from Greenwich.

alexandercyu committed:
Update node-login.css

Add style for timezone dropdown so vertical scrollbar is added for long lists

alexandercyu committed:
Update dropdown-publish.html

- Add timezone dropdown version to the template
- reuse the UL tag of the dropdown list in the timezone dropdown, but add the style for overflow-y auto
- list items for timezone dropdown is all built in javascript instead of jinja

alexandercyu committed:
Update publish-dataproduct.html

Add timezone dropdown to the data product page

alexandercyu committed:
Update dropdown-publish.html

Enable addTimezoneToSelectedTime() function here for testing

alexandercyu committed:
Update dataproduct-form.js

Add function to convert the chosen dates from the calendar to a UTC timestamp without the timezone, and  include the offset for the user chosen timezone

alexandercyu committed:
Update dataproduct-retrieve.js

Test filling out user chosen calendar dates and timezone using server data

alexandercyu committed:
Update dropdown-publish.html

- change timezone dropdown so it's only UTC and GMT timezone identifiers
- add the timezone offset to the anchor tag to be used when filling out the form again with server data

alexandercyu committed:
Update dataproduct-form.js

Add function to generate list of timezones with identifiers and offsets to build the timezone dropdown

alexandercyu committed:
Update dataproduct-retrieve.js

- parse out the timezone offset from the server temporal value and select the appropriate item in the timezone dropdown list

alexandercyu committed:
Update publish-dataproduct.html

Change ID of timezone dropdown

mishaschwartz commented on templates/partials/dropdown-publish.html:
```diff
@@ -86,6 +90,66 @@ <h6 id="{{ dropdown_button_text_id }}" class="dropdown-list-title margin-unset p
+                        generatedTimezonesList.forEach( (generatedTimezoneListItem, generatedTimezoneListIndex) => {
+                            stickiedTimezoneList.forEach( (stickiedTimezone, stickiedTimezoneIndex) => {
+                                if(generatedTimezoneListItem.id === stickiedTimezone || generatedTimezoneListItem.id.includes(stickiedTimezone)){
+                                    timezonesWithOffsetsList.splice(0,0, generatedTimezoneListItem);
```
Why `splice`? Why not just `push`? Does it matter to add it to the front of the list?

mishaschwartz replied:
This misses out a lot of valid offsets that don't fall exactly on the hour.

All of these would be missing:

America/St_Johns
Asia/Colombo
Asia/Kabul
Asia/Kathmandu
Asia/Kolkata
Asia/Tehran
Asia/Yangon
Australia/Adelaide
Australia/Broken_Hill
Australia/Darwin
Australia/Eucla
Indian/Cocos
Pacific/Chatham
Pacific/Marquesas

alexandercyu replied:
Changed how timezone list is made to include all timezones offsets instead of just the whole hour offsets.

mishaschwartz commented on templates/site/js/dataproduct-form.js:
```diff
@@ -1359,14 +1431,20 @@ async function submitForm(){
+        //var dateISOString = dateUTC.toISOString();
+        submitObject["temporal"].push(dateMetadata);
     }
+    console.log()
```
remove debugging

mishaschwartz reviewed (changes requested):
Some additional issues:

- the -/+ sign is the opposite of what is sent to the backend (if I choose a +2 offset the value in the backend is -2)
- there's a lot of commented code and console.log lines still in here. I've commented on a few but there's lots

alexandercyu committed:
Update dropdown-publish.html

Build timezone dropdown using all timezone offsets instead of just whole hour offsets

alexandercyu committed:
Update dataproduct-form.js

- Format the time using the timezone offset instead of the identifier name
- remove commented code
- remove unneeded code

alexandercyu committed:
Update dataproduct-retrieve.js

Get the full offset name instead of just the time part

alexandercyu committed:
Update dropdown-list.js

Add the selected_offset value to the dropdown label

mishaschwartz commented on templates/site/js/dataproduct-form.js:
```diff
@@ -893,7 +893,76 @@ function showHideModelInput(dropdownItemIndex, dropdownItemName, dropdownIndex){
+    var formattedStartDate = dateFormatter.format(startDate);
+    var formattedEndDate = dateFormatter.format(endDate);
+
+    var startDateSplitTimezoneArray = formattedStartDate.split("GMT");
```
Use the same code you did before instead of splitting on parts of the string: 

```
        const offsetPart = parts.find(part => part.type === 'timeZoneName');
```

alexandercyu replied:
The formatToParts function returns the timezone with "GMT" in front.  If I just want the hour value I still have to split.

mishaschwartz commented on templates/site/js/dataproduct-retrieve.js:
```diff
@@ -151,20 +152,41 @@ async function populateForm(){
+                var offsetIndicator = serverStartDate.charAt(indexOfLastColon - 3);
+                var serverTimezoneArray = serverStartDate.split(offsetIndicator);
+
+                if(offsetIndicator == "-"){
```
This is very hacky and counterintuitive. You already have the index of the offsetIndicator. Just use that instead of splitting and having to work around the fact that hyphens and minus signs use the same character.

alexandercyu replied:
Changed this to use the index of the offsetIndicator instead of searching for the offsetIndicator character.

mishaschwartz commented on templates/site/js/dataproduct-retrieve.js:
```diff
@@ -151,20 +152,41 @@ async function populateForm(){
 
-                metadataStartDate._flatpickr.setDate(firstDateString);
-                metadataEndDate._flatpickr.setDate(secondDateString);
+                metadataStartDate._flatpickr.setDate(serverStartDate.replace("T", " "));
```
Why remove the T. Flatpicker can handle iso date strings

mishaschwartz commented on templates/site/js/dropdown-list.js:
```diff
@@ -3,10 +3,16 @@ function replaceListItem(titleItemID, selectedItemID){
     firstItem.setAttribute("selected_index", selectedIndex);
     firstItem.setAttribute("selected_id", selectedID);
+
+    if(secondItem.hasAttribute("selected_offset")){
```
custom attributes in html must have the `data-` prefix

alexandercyu committed:
Update dropdown-publish.html

- Change custom attribute names for "selected_offset" and "selected_index" to include the "data-" prefix.  Other attribute names will be changed in a different PR because changing them now caused errors in HTML validation
- move functionality for generating timezone dropdown list items to its own function in dropdown-list.js

alexandercyu committed:
Update dataproduct-form.js

- Change custom attribute names for "selected_offset" and "selected_index" to include the "data-" prefix.  Other attribute names will be changed in a different PR because changing them now caused errors in HTML validation
- removed unused formatter options
- change format() to formatToParts() to use the find() to get the timezone offset value

alexandercyu committed:
Update dropdown-publish.html

- add parameters to generateTImezoneDropdownListItems()

alexandercyu committed:
Update dataproduct-form.js

-Add condition to handle case of GMT offset (+00:00)
-> Will keep "GMT" label of timezone dropdown as "GMT" while the attribute "selected_index" will be "GMT+00:00"

alexandercyu committed:
Update dataproduct-retrieve.js

- add condition to handle when timezone includes "Z" or "+00:00" so it will show "GMT" in the dropdown
- change custom attributes to include "data-" prefix except for "selected_value" because changing "selected_value" will cause a HTML validation error
- remove the lines replacing "T" in the iso timezone  format

alexandercyu committed:
Update dropdown-list.js

- add "data-" prefix to custom attributes except for "selected_value"
- move function for building timezone dropdown list items here from the template

alexandercyu committed:
Update node-login.css

Add styles for timezone dropdown invalid status

alexandercyu committed:
Update dropdown-publish.html

Add containerID to the parameters

alexandercyu committed:
Update dataproduct-form.js

- Add code to set a custom validity message on a hidden input field in the timezone dropdown when there is no timezone value chosen, and remove the message when there is

alexandercyu committed:
Update dropdown-list.js

- add code to add an input field w with no height or width to the timezone dropdown to be used to set a custom validity message during HTML validation

alexandercyu committed:
Update publish-dataproduct.html

Remove timezone  description for temporal input section
